src/aiida_quantumespresso/workflows/ph/base.py

```diff
@@ -45,6 +56,8 @@ def define(cls, spec):
             cls.results,
         )
         spec.expose_outputs(PhCalculation, exclude=('retrieved_folder',))
+        spec.exit_code(400, 'ERROR_INVALID_INPUT_QPOINTS',
```

Comment:
Why can't we already raise the error before the process is started? E.g. via a `validator`?

Reply:
Hi @bastonero , could you put an example ?

Reply:
You can add after the input specifications
```python
spec.inputs.validator = validate_inputs
```
and then outside the class something like this:
```python
def validate_inputs(inputs, _):
    """Validate the top level namespace."""
    if 'qpoints_distance' not in inputs and 'qpoints' not in inputs:
            return 'Either `qpoints_distance` or `qpoints` must be given`
```
And then it would be great if you can write a simple test (you can look into the tests folder and look for something similar in the validations of the inputs)

Reply:
This way you don't start the process from the very beginning avoiding to store a node that would fail due to "bad inputs"

Reply:
Do we really need the exit code here?

Reply:
See also the validation function comment, but essentially I would remove this exit code. It's not useful as it is only a validation error